packages/tpetra/core/src/Tpetra_Experimental_BlockCrsMatrix_def.hpp

```diff
@@ -2835,91 +2928,96 @@ class GetLocalDiagCopy {
     // have different numbers of entries.
     constantNumPackets = 0;
 
+    // const values
+    const crs_graph_type& srcGraph = src->graph_;
+    const size_t blockSize = static_cast<size_t> (src->getBlockSize ());
+    const size_t numExportLIDs = exportLIDs.extent (0);
+    // Kyungjoo: We do not consider runtime sizable types e.g., Stokhos DFAD
```

Comment:
@etphipp Do we have a function to check the type ? 
@mhoemmen Do we have an explicit dependence on SACADO or STOKHOS so that we can put ifdef guard around the check the variable ? 

If nobody uses DFAD with Tpetra, I personally think that we must prohibit from using DFAD with Tpetra. There is no real reason to use DFAD. 

1. If we allow DFAD and a value may have a different derivative dimension, we need to check byte per value for each entry value. This is not feasible.
2. If DFAD or UQ types are only used with fixed subarray size, then there is no reason to use DFAD (use SFAD).

How do you think ?

Reply:
DFad, or Sacado in general, is not currently used with Tpetra.  Stokhos's UQ scalar types (Sacado::UQ::PCE and Sacado::MP::Vector) are used with Tpetra however (but not currently with BlockCRS).  The PCE scalar type is dynamically sized, however there were already methods in Tpetra for handling this.

Reply:
It is true that they are all the same size.  There is a way to get the size without accessing values inside the View, but the problem is the function is only available in Stokhos/Sacado currently (and Kokkos has resisted adding any function like it to Kokkos itself).  Can't you just check if the array has zero size first?

Reply:
Although I can just check the array size for now, I am just thinking about a more clear error triggering when someone attempt to use a runtime sizable type and encouters such a unfortunate edge case. It should be better than getting incorrect results. Maybe I worry too much. Okay, let's just check the array size and get this PR merged.

Reply:
@etphipp wrote:
> There is a way to get the size without accessing values inside the View....

Is there a way to get the size by just knowing the Scalar type, without needing a Scalar instance?  If so, could we just put that in the Stokhos/Sacado specializations of PackTraits?

Useful Scalar types do exist where each instance of Scalar may have a different run-time size, but Tpetra can't handle that case now anyway.  (MultiVector Ex/Import would need two rounds of communication in that case.)

Reply:
> Is there a way to get the size by just knowing the Scalar type, without needing a Scalar instance?

Yes, for a given View v, just do `Kokkos::dimension_scalar(v)`.